docs: platform v12 release blog (#167)
dhis2/developer-portal: blog/2024-11-2-app-platform-v12.md

---
slug: 2024/11/app-platform-v12
title: App Platform v12

> authors: core

Reviewer: I recommend adding your own name :) 
```suggestion
authors: [core, kai]
```

tags: [app platform, developer tools, webapp, announcement]
---

> # Platform v12 Migration Guide: Vite & React 18

Reviewer: Putting `release` first as this is a release announcement, but also has migration included, so putting that secondary. 

```suggestion
# Platform v12 release, Vite and React 18 Migration Guide
```

> Good news! Vite and React 18 in the app platform are ready to use!

Reviewer: ```suggestion
Good news! With the App Platform v12 release, we've replaced the deprecated Create React App (CRA) with Vite, and upgraded React from v16 to v18!
```

Author: I edited this a little bit: https://github.com/dhis2/developer-portal/pull/167/commits/ff2cb6fb5d4f181b6733b08451ad71d9b0b18ef7 -- let me know what you think

Reviewer: looks good!

> We're very excited for these updates — both will significantly modernize the App Platform, and Vite will be a big upgrade from Create React App (CRA), which we used under the hood to build and serve apps previously. It will greatly improve the developer experience, and with greater control over the configuration, it will open up some powerful new possibilities for the platform.

Reviewer: I'm suggesting this, as my previous suggestion already explained CRA. 

```suggestion
We're very excited for these updates — both will significantly modernize the App Platform, and Vite will be a big upgrade from CRA, which we used under the hood to build and serve apps previously. It will greatly improve the developer experience, and with greater control over the configuration, it will open up some powerful new possibilities for the platform.
```

    -   Code is shared between entrypoints, which makes bundles smaller
    -   Hot Module Replacement (HMR) for code changes in the plugin will be as fast as in the app
-   There are some new TypeScript features:
    -   Bootstrapping an app with a TypeScript template is supported: `d2-app-scripts init --typescript my-app`. See the [`init` docs](https://example.org)<!-- todo: (app-platform/scripts/init.md) --> for more about the command
    -   Vite has native support for TypeScript
    -   Although not in the App Platform package, with the latest `@dhis2/cli-style`, TypeScript type checking is performed when running `d2-style lint`
    -   With `@dhis2/ui` version 9, the UI library is now typed as well
-   There's a small suite of tools available in the CLI when running an app in dev mode:
    -   With the dev server running, press `h + enter` to see the options
    -   Options include exposing the server on LAN, opening the app in the browser, cleanly quitting the server, and restarting the dev server (which can be helpful if you're modifying libraries in `node_modules`)
-   The build output includes a summary to inspect chunks

### Future changes

> With Vite, the door is open for some big future improvements:

Reviewer: ```suggestion
With Vite, the door is open for some big future improvements. We've already created JIRA tickets for these, so if you're interested in staying up-to-date on them you can watch these tickets.
```

-   Extensible config: developers can add their own options to the Vite config, for example a plugin for Flow types, or to define import aliases ([LIBS-706](https://dhis2.atlassian.net/browse/LIBS-706))
-   Arbitrary entrypoints, beyond app/plugin/lib: Make a regular app, a configuration app, a capture plugin, a dashboard plugin, and more all from the same repo and sharing code between them ([LIBS-394](https://dhis2.atlassian.net/browse/LIBS-394))

With respect to TypeScript, work is also under way to add types to data fetching tools using specs generated from OpenAPI ([LIBS-523](https://dhis2.atlassian.net/browse/LIBS-523)).

## Getting started

By running these steps, you should be able to run your app right away:

1. `yarn add @dhis2/app-runtime @dhis2/ui -D @dhis2/cli-app-scripts`
2. `npx yarn-deduplicate yarn.lock && yarn`
3. Try out `yarn start --allowJsxInJs`, and your app should be running 🚀

> There will be some other changes you will want to make, which are described in detail in the [migration guide](https://example.org)<!-- todo: (app-platform/migration/v12) -->. Our goal is to make it easy to adopt the new changes, so we have some tools to facilitate the process.

Reviewer: ```suggestion
There will be some other changes you will want to make, which are described in detail in the [migration guide](https://example.org)<!-- todo: (/docs/app-platform/migration/v12) -->. Our goal is to make it easy to adopt the new changes, so we have some tools to facilitate the process.
```

> Head on over to the [migration guide](https://example.org)<!-- todo: (app-platform/migration/v12) --> for more instructions and technical info.

Reviewer: ```suggestion
Head on over to the [migration guide](https://example.org)<!-- todo: (/docs/app-platform/migration/v12) --> for more detailed instructions and technical background information.
```
